README.md
# Error Kit

Error Kit is an NPM package that provides a collection of common custom error classes for TypeScript. These error classes are designed to extend TypeScript errors and provide additional functionality and enhancements for error handling in your applications.

### Installation

Install the package using npm:

```shell
npm install @imraan-hendricks/error-kit
```

or with Yarn:

```shell
yarn add @imraan-hendricks/error-kit
```

### Usage

To use the custom error classes from Error Kit in your TypeScript/Node.js application, you can import the specific error classes and create instances as needed.

### Example

```typescript
import { ForbiddenError, NotFoundError } from '@imraan-hendricks/error-kit';

try {
  // Perform some operations that may throw errors
  throw new ForbiddenError('Access denied.');
} catch (error) {
  if (error instanceof ForbiddenError) {
    console.log('Forbidden error occurred:', error.message);
    console.log('Status Code:', ForbiddenError.statusCode);
    // Handle the forbidden error
  } else if (error instanceof NotFoundError) {
    console.log('Not found error occurred:', error.message);
    console.log('Status Code:', NotFoundError.statusCode);
    // Handle the not found error
  } else {
    console.log('An error occurred:', error.message);
    // Handle other types of errors
  }
}
```

### Generic Error

The GenericError class is a special class that extends the built-in Error class, serving as a base error class for both generic errors that do not fall into a specific category and other custom error classes. All custom error classes provided by Error Kit are instances of the GenericError class, ensuring consistent behavior and functionality.

```typescript
import {
  GenericError,
  NotFoundError,
  ValidationError,
  InternalServerError,
  UnauthorizedError,
} from '@imraan-hendricks/error-kit';

const notFoundError = new NotFoundError('Resource not found.');
const internalServerError = new InternalServerError('Internal server error.');
const unauthorizedError = new UnauthorizedError('Unauthorized access.');

console.log(notFoundError instanceof GenericError); // true
console.log(internalServerError instanceof GenericError); // true
console.log(unauthorizedError instanceof GenericError); // true
```

### Validation Error

The ValidationError class is a unique error class that allows you to pass a generic type parameter <T> to define the type of the errors value. It represents a validation error and includes an array of validation errors.

```typescript
import { ValidationError } from '@imraan-hendricks/error-kit';

// Define a custom type for validation errors
type MyValidationError = {
  field: string;
  message: string;
};

// Create an array of validation errors
const validationErrors: MyValidationError[] = [
  { field: 'name', message: 'Name is required.' },
  { field: 'email', message: 'Invalid email format.' },
];

// Create a new instance of ValidationError with the custom error type
const validationError = new ValidationError<MyValidationError>(
  validationErrors,
  'Validation failed.',
);

console.log(validationError.message);
console.log(validationError.errors);
```

By using TypeScript generics with the ValidationError class, you can customize the type of the errors value to match the specific structure and format of your validation errors.

### Sending Errors as HTTP Responses

You can easily send custom errors as HTTP responses using the toJson() function and res.json() method provided by popular Node.js frameworks. The toJson() function converts the custom error object into a plain JavaScript object that can be serialized as JSON.

```typescript
import express from 'express';
import { NotFoundError } from '@imraan-hendricks/error-kit';

const app = express();

app.get('/not-found', (req, res) => {
  const error = new NotFoundError('Resource not found.');
  const statusCode = NotFoundError.statusCode;
  res.status(statusCode).json(error.toJson());
});

app.listen(3000, () => {
  console.log('Server is running on port 3000');
});
```

### Available Error Classes

The package includes the following custom error classes:

- BadRequestError
- DatabaseError
- ForbiddenError
- GenericError
- InternalServerError
- NotAcceptableError
- NotFoundError
- NoRecordError
- ServiceUnavailableError
- UnauthorizedError
- ValidationError

Each error class extends the base TypeScript Error class and provides additional functionality and enhancements for error handling. You can override the default message by passing a custom message to the constructor.

For more detailed information, advanced usage, and customization options, please refer to the [documentation](https://github.com/Imraan-Hendricks/error-kit).

### Contributing

Contributions are always welcome! If you encounter any issues, have suggestions, or want to contribute improvements, please open an issue or submit a pull request in the [Github repository](https://github.com/Imraan-Hendricks/error-kit).

### Support

If you have any questions, issues, or need assistance, please feel free to reach out to our support team at support@imraanhendricks.com. We are here to help and will get back to you as soon as possible.

### License

This project is licensed under the [MIT License](https://github.com/Imraan-Hendricks/error-kit/blob/main/LICENSE).


changelog.md
## 6/10/2023: v0.1.0

### Features

- add NotAcceptableError class for unsupported content types
- add NotFoundError class for handling missing resources
- add NoRecordError class for handling missing records
- add ServiceUnavailableError class for handling service unavailability
- add UnauthorizedError class for handling unauthorized access
- add ValidationError class for handling validation errors
- add BadRequestError for handling invalid requests
- add ForbiddenError class for handling forbidden access
- introduce GenericError class for custom errors

### Chores

- set appropriate keywords in package.json for better discoverability
- create necessary configuration files for linting and formatting code

### Documentation

- add a comprehensive README.md file with usage instructions and examples
- document the purpose, usage, and features of each error class
- add appropriate license information to the package
- include necessary metadata such as homepage, bugs, and keywords in package.json

### Builds

- configure Rollup for bundling the project

### CI

- create CI/CD workflow for automated builds and tests


src/index.spec.ts
import { GenericError } from './';

describe('main', () => {
  it('should correctly export named errors', () => {
    const error = new GenericError();
    expect(error instanceof Error).toBe(true);
  });
});


src/utils/error-utils.spec.ts
import {
  BadRequestError,
  DatabaseError,
  ForbiddenError,
  GenericError,
  InternalServerError,
  NoRecordError,
  NotAcceptableError,
  NotFoundError,
  ServiceUnavailableError,
  UnauthorizedError,
  ValidationError,
} from './error-utils';

describe('Error utilities', () => {
  describe('GenericError', () => {
    it('should be an instance of Error', () => {
      const error = new GenericError();
      expect(error instanceof Error).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new GenericError();
      expect(error.message).toBe('An error occurred.');
      expect(error.name).toBe('GenericError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new GenericError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new GenericError();
      expect(error.statusCode).toBe(500);
      expect(GenericError.statusCode).toBe(500);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new GenericError(errorMessage);
      const expectedJson = {
        statusCode: GenericError.statusCode,
        name: 'GenericError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('BadRequestError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new BadRequestError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new BadRequestError();
      expect(error.message).toBe('The request contains invalid data.');
      expect(error.name).toBe('BadRequestError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new BadRequestError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new BadRequestError();
      expect(error.statusCode).toBe(400);
      expect(BadRequestError.statusCode).toBe(400);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new BadRequestError(errorMessage);
      const expectedJson = {
        statusCode: BadRequestError.statusCode,
        name: 'BadRequestError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('DatabaseError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new DatabaseError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new DatabaseError();
      expect(error.message).toBe(
        'An error occurred while accessing the database.',
      );
      expect(error.name).toBe('DatabaseError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new DatabaseError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new DatabaseError();
      expect(error.statusCode).toBe(500);
      expect(DatabaseError.statusCode).toBe(500);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new DatabaseError(errorMessage);
      const expectedJson = {
        statusCode: DatabaseError.statusCode,
        name: 'DatabaseError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('ForbiddenError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new ForbiddenError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new ForbiddenError();
      expect(error.message).toBe('Access forbidden.');
      expect(error.name).toBe('ForbiddenError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new ForbiddenError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new ForbiddenError();
      expect(error.statusCode).toBe(403);
      expect(ForbiddenError.statusCode).toBe(403);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new ForbiddenError(errorMessage);
      const expectedJson = {
        statusCode: ForbiddenError.statusCode,
        name: 'ForbiddenError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('InternalServerError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new InternalServerError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new InternalServerError();
      expect(error.message).toBe('An unexpected error occurred on the server.');
      expect(error.name).toBe('InternalServerError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new InternalServerError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new InternalServerError();
      expect(error.statusCode).toBe(500);
      expect(InternalServerError.statusCode).toBe(500);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new InternalServerError(errorMessage);
      const expectedJson = {
        statusCode: InternalServerError.statusCode,
        name: 'InternalServerError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('NotAcceptableError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new NotAcceptableError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new NotAcceptableError();
      expect(error.message).toBe(
        'The requested content type is not supported by the server.',
      );
      expect(error.name).toBe('NotAcceptableError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new NotAcceptableError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new NotAcceptableError();
      expect(error.statusCode).toBe(406);
      expect(NotAcceptableError.statusCode).toBe(406);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new NotAcceptableError(errorMessage);
      const expectedJson = {
        statusCode: NotAcceptableError.statusCode,
        name: 'NotAcceptableError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('NotFoundError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new NotFoundError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new NotFoundError();
      expect(error.message).toBe('The requested resource was not found.');
      expect(error.name).toBe('NotFoundError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new NotFoundError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new NotFoundError();
      expect(error.statusCode).toBe(404);
      expect(NotFoundError.statusCode).toBe(404);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new NotFoundError(errorMessage);
      const expectedJson = {
        statusCode: NotFoundError.statusCode,
        name: 'NotFoundError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('NoRecordError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new NoRecordError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new NoRecordError();
      expect(error.message).toBe('The requested record does not exist.');
      expect(error.name).toBe('NoRecordError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new NoRecordError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new NoRecordError();
      expect(error.statusCode).toBe(404);
      expect(NoRecordError.statusCode).toBe(404);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new NoRecordError(errorMessage);
      const expectedJson = {
        statusCode: NoRecordError.statusCode,
        name: 'NoRecordError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('ServiceUnavailableError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new ServiceUnavailableError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new ServiceUnavailableError();
      expect(error.message).toBe(
        'Service is currently unavailable. Please try again later.',
      );
      expect(error.name).toBe('ServiceUnavailableError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new ServiceUnavailableError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new ServiceUnavailableError();
      expect(error.statusCode).toBe(503);
      expect(ServiceUnavailableError.statusCode).toBe(503);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new ServiceUnavailableError(errorMessage);
      const expectedJson = {
        statusCode: ServiceUnavailableError.statusCode,
        name: 'ServiceUnavailableError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('UnauthorizedError', () => {
    it('should be an instance of Error and GenericError', () => {
      const error = new UnauthorizedError();
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new UnauthorizedError();
      expect(error.message).toBe('Access denied. Authorization required.');
      expect(error.name).toBe('UnauthorizedError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new UnauthorizedError(errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should have the correct statusCode', () => {
      const error = new UnauthorizedError();
      expect(error.statusCode).toBe(401);
      expect(UnauthorizedError.statusCode).toBe(401);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new UnauthorizedError(errorMessage);
      const expectedJson = {
        statusCode: UnauthorizedError.statusCode,
        name: 'UnauthorizedError',
        message: errorMessage,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });

  describe('ValidationError', () => {
    const errors = [{ name: 'customName', message: 'custom message' }];

    it('should be an instance of Error and GenericError', () => {
      const error = new ValidationError(errors);
      expect(error instanceof Error).toBe(true);
      expect(error instanceof GenericError).toBe(true);
    });

    it('should set the default message and name correctly', () => {
      const error = new ValidationError(errors);
      expect(error.message).toBe(
        'The provided data does not meet the required criteria.',
      );
      expect(error.name).toBe('ValidationError');
    });

    it('should set the custom message correctly', () => {
      const errorMessage = 'Custom error message';
      const error = new ValidationError(errors, errorMessage);
      expect(error.message).toBe(errorMessage);
    });

    it('should set the errors correctly', () => {
      const error = new ValidationError(errors);
      expect(error.errors).toEqual(errors);
    });

    it('should have the correct statusCode', () => {
      const error = new ValidationError(errors);
      expect(error.statusCode).toBe(400);
      expect(ValidationError.statusCode).toBe(400);
    });

    it('should return the correct JSON representation', () => {
      const errorMessage = 'Error message';
      const error = new ValidationError(errors, errorMessage);
      const expectedJson = {
        statusCode: ValidationError.statusCode,
        name: 'ValidationError',
        message: errorMessage,
        errors,
      };
      expect(error.toJSON()).toEqual(expectedJson);
    });
  });
});


src/utils/error-utils.ts
export class GenericError extends Error {
  static statusCode: number = 500;
  public statusCode: number = GenericError.statusCode;
  public name: string = 'GenericError';

  constructor(message: string = 'An error occurred.') {
    super(message);
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, GenericError);
    }
  }

  toJSON() {
    return {
      statusCode: this.statusCode,
      name: this.name,
      message: this.message,
    };
  }
}

export class BadRequestError extends GenericError {
  static statusCode: number = 400;
  public statusCode: number = BadRequestError.statusCode;
  public name: string = 'BadRequestError';

  constructor(message: string = 'The request contains invalid data.') {
    super(message);
  }
}

export class DatabaseError extends GenericError {
  static statusCode: number = 500;
  public statusCode: number = DatabaseError.statusCode;
  public name: string = 'DatabaseError';

  constructor(
    message: string = 'An error occurred while accessing the database.',
  ) {
    super(message);
  }
}

export class ForbiddenError extends GenericError {
  static statusCode: number = 403;
  public statusCode: number = ForbiddenError.statusCode;
  public name: string = 'ForbiddenError';

  constructor(message: string = 'Access forbidden.') {
    super(message);
  }
}

export class InternalServerError extends GenericError {
  static statusCode: number = 500;
  public statusCode: number = InternalServerError.statusCode;
  public name: string = 'InternalServerError';

  constructor(message: string = 'An unexpected error occurred on the server.') {
    super(message);
  }
}

export class NotAcceptableError extends GenericError {
  static statusCode: number = 406;
  public statusCode: number = NotAcceptableError.statusCode;
  public name: string = 'NotAcceptableError';

  constructor(
    message: string = 'The requested content type is not supported by the server.',
  ) {
    super(message);
  }
}

export class NotFoundError extends GenericError {
  static statusCode: number = 404;
  public statusCode: number = NotFoundError.statusCode;
  public name: string = 'NotFoundError';

  constructor(message: string = 'The requested resource was not found.') {
    super(message);
  }
}

export class NoRecordError extends GenericError {
  static statusCode: number = 404;
  public statusCode: number = NoRecordError.statusCode;
  public name: string = 'NoRecordError';

  constructor(message: string = 'The requested record does not exist.') {
    super(message);
  }
}

export class ServiceUnavailableError extends GenericError {
  static statusCode: number = 503;
  public statusCode: number = ServiceUnavailableError.statusCode;
  public name: string = 'ServiceUnavailableError';

  constructor(
    message: string = 'Service is currently unavailable. Please try again later.',
  ) {
    super(message);
  }
}

export class UnauthorizedError extends GenericError {
  static statusCode: number = 401;
  public statusCode: number = UnauthorizedError.statusCode;
  public name: string = 'UnauthorizedError';

  constructor(message: string = 'Access denied. Authorization required.') {
    super(message);
  }
}

export class ValidationError<T> extends GenericError {
  static statusCode: number = 400;
  public statusCode: number = ValidationError.statusCode;
  public name: string = 'ValidationError';
  public errors: T[];

  constructor(
    errors: T[],
    message: string = 'The provided data does not meet the required criteria.',
  ) {
    super(message);
    this.errors = errors;
  }

  toJSON() {
    return {
      statusCode: ValidationError.statusCode,
      name: this.name,
      message: this.message,
      errors: this.errors,
    };
  }
}
